src/test/java/io/cryostat/configuration/CredentialsManagerTest.java

```diff
@@ -0,0 +1,540 @@
+                            "%s.json",
+                            base32.encodeToString(
+                                    matchExpression.getBytes(StandardCharsets.UTF_8)));
+
+            Path originalPath = Mockito.mock(Path.class);
+            Path newPath = Mockito.mock(Path.class);
+            Mockito.when(credentialsDir.resolve(originalFilename)).thenReturn(originalPath);
+            Mockito.when(credentialsDir.resolve(newFilename)).thenReturn(newPath);
+
+            String username = "user";
+            String password = "pass";
+
+            String oldContents =
+                    gson.toJson(
+                            Map.of(
+                                    "targetId",
+                                    targetId,
+                                    "credentials",
+                                    Map.of("username", username, "password", password)));
+            Mockito.when(fs.readFile(originalPath))
+                    .thenReturn(new BufferedReader(new StringReader(oldContents)));
+
+            credentialsManager.migrate();
+
+            ArgumentCaptor<String> contentsCaptor = ArgumentCaptor.forClass(String.class);
+
+            Mockito.verify(fs).listDirectoryChildren(credentialsDir);
+            Mockito.verify(credentialsDir).resolve(originalFilename);
+            Mockito.verify(fs).readFile(originalPath);
+            Mockito.verify(fs)
+                    .writeString(
+                            Mockito.eq(newPath),
+                            contentsCaptor.capture(),
+                            Mockito.eq(StandardOpenOption.WRITE),
+                            Mockito.eq(StandardOpenOption.CREATE),
+                            Mockito.eq(StandardOpenOption.TRUNCATE_EXISTING));
+            Mockito.verify(fs)
+                    .setPosixFilePermissions(
+                            newPath,
+                            Set.of(
+                                    PosixFilePermission.OWNER_READ,
+                                    PosixFilePermission.OWNER_WRITE));
+            Mockito.verify(fs).deleteIfExists(originalPath);
+            Mockito.verifyNoMoreInteractions(fs);
+
+            String newContents = contentsCaptor.getValue();
+            JsonObject json = gson.fromJson(newContents, JsonObject.class);
+            MatcherAssert.assertThat(
+                    json.getAsJsonPrimitive("matchExpression").getAsString(),
+                    Matchers.equalTo(matchExpression));
+            JsonObject credentials = json.getAsJsonObject("credentials");
+            String foundUsername = credentials.getAsJsonPrimitive("username").getAsString();
+            String foundPassword = credentials.getAsJsonPrimitive("password").getAsString();
+            MatcherAssert.assertThat(foundUsername, Matchers.equalTo(username));
+            MatcherAssert.assertThat(foundPassword, Matchers.equalTo(password));
+        }
+    }
+
+    @Nested
+    class Loading {
```

Comment:
Could you explain the difference between these two tests? I noticed that the second one doesn't require `Mockito.when(platformClient.listDiscoverableServices()).thenReturn(List.of(serviceRef));` even though they seem functionally identical.

Reply:
Between the two `Loading` tests? One calls `credentialsManager.getCredentials(ServiceRef serviceRef)` and the other calls `credentialsManager.getCredentialsByTargetId(String targetId)`. The `byTargetId` one is there to preserve backward-compatibility with the older API and the `AbstractAuthenticatedRequestHandler`, so that the new style can be used for API paths like `/api/v1/targets/:targetId/recordings`, where we don't have a full `ServiceRef` object to evaluate the `matchExpression` against. In that case, the credentials manager searches for known targets that match the given `targetId`, maps that to a `ServiceRef` if it can, and then checks for matching credentials for that. The `getCredentials(ServiceRef)` just directly looks for credentials that match the supplied `ServiceRef`, without having to call out to the platform client to query for targets that match (since this is already a reference to a target).

Reply:
Makes sense, thanks. I should've just diffed the two files instead of looking to see what they differences were; missed the `getCredentials(...)` vs. `getCredentialsByTargetId(...)` calls :P